src/software/power/charger.cpp

```diff
@@ -0,0 +1,45 @@
+#include "charger.h"
+
+volatile bool Charger::flyback_fault    = false;
+void (*Charger::charge_done_callback)() = NULL;
+
+Charger::Charger()
+{
+    pinMode(HV_SENSE, INPUT);
+    pinMode(FLYBACK_FAULT, INPUT);
+    pinMode(CHRG_DONE, INPUT);
+    pinMode(CHRG, OUTPUT);
+
+    attachInterrupt(digitalPinToInterrupt(CHRG_DONE), chargeDone, RISING);
+}
+
+void Charger::chargeDone()
+{
+    flyback_fault = digitalRead(FLYBACK_FAULT);
+    if (charge_done_callback)
+    {
+        charge_done_callback();
+        charge_done_callback = NULL;
+        setChargeMode(LOW);
+    }
+}
+
+void Charger::setChargeMode(int op)
+{
+    digitalWrite(CHRG, op);
+}
+
+void Charger::setChargeDoneCallbackOnce(void (*charge_done_callback)())
+{
+    Charger::charge_done_callback = charge_done_callback;
+}
+
+float Charger::getCapacitorVoltage()
+{
+    return analogRead(HV_SENSE);
```

Comment:
The ADC resolution that the esp reads is 12 bit. This means the max value analogread() will output is 2^12 = 4096, which would correspond to a voltage of 3.2-3.3V (ESP32 has trouble distinguishing between 3.2-3.3V and 0-0.1V - not a problem for us tho because the max voltage we should read is about 3V when caps are charged to 238V). When you call analogread, it will give a value between 0 and 4096, and to get the actual voltage you need to do the following: Actual voltage = output from AnalogRead() / 4096 * 3.3V